# Assignment 8: Dimensionality Reduction

Dataset(s) needed: MNIST ("Modified National Institute of Standards and Technology") dataset.

# Learning Objectives
- Be able to make application decisions regarding principal component analysis to train and test data 
- Produce a dimensionality reduction model.

In [30]:
#Load the MNIST dataset
from sklearn.datasets import fetch_openml
#mnist = fetch_openml('mnist_784')

# Get the X and y samples
X, y = fetch_openml('mnist_784', version=1, return_X_y=True)
print(X.shape)
print(y.shape)

# Normalize the X samples to be between 0 and 1
X = (X/255).astype('float32')

X.head()


C:\Users\louis\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\datasets\_openml.py:1022: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


(70000, 784)
(70000,)


pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  pixel9  \
0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   pixel10  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
1      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
2      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
3      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   
4      0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   

   pixel781  pixel782  pixel783  pixel784  
0       0.0       0.0       0.0       0.0  
1       0.0       0.0       0.0       0.0  
2       0.0       0.0       0.0       0.0  
3       0.0       0.0       0.0       0.0  
4       0.0       0.0       0.0       0.0  

[5 rows x 784 columns]

Question 1: Split the data into a training set and a test set (take the first 60,000 instances for training, and the remaining 10,000 for testing).

In [31]:
# Split the samples into a training set and a test set
X_train = X[:60000]
Y_train = y[:60000]

X_test = X[60000:]
Y_test = y[60000:]



Question 2: Train a Logistic Regression classifier on the dataset. Set solver = lbfgs and random_state = 42. Hint you might also want to think about the multi_class parameter and if you need to set that.  You want to track how long the model takes to run.

In [32]:
from sklearn.linear_model import LogisticRegression
import time

time_start = time.time()
logistic_regression_model = LogisticRegression(solver='lbfgs', max_iter=5000, random_state=42, verbose=1)
logistic_regression_model.fit(X_train, Y_train)
print('Logistic regression training done! Time elapsed: {} seconds'.format(time.time()-time_start))

Logistic regression training done! Time elapsed: 32.47798013687134 seconds


Question 3: Evaluate the resulting model on the test set.

In [33]:
# Use score method to get accuracy of model
score = logistic_regression_model.score(X_test, Y_test)
print(score)


0.9256


Question 4: Use PCA to reduce the dataset's dimensionality, with an explained variance ratio of 95%.

In [34]:
from sklearn.decomposition import PCA
time_start = time.time()
pca = PCA(n_components=0.95)
X_pca = pca.fit_transform(X)


X_train_pca = X_pca[:60000]
X_test_pca = X_pca[60000:]
print('PCA analysis done! Time elapsed: {} seconds'.format(time.time()-time_start))

PCA analysis done! Time elapsed: 2.3028647899627686 seconds


Question 5: Train a new Logistic Regression classifier on the reduced dataset and see how long it takes. Use the same criteria as before just with PCA. Was training much faster? Explain your results

In [35]:

logistic_regression_model = LogisticRegression(solver='lbfgs', max_iter=5000, random_state=42, verbose=1)
logistic_regression_model.fit(X_train_pca, Y_train)
print('Logistic regression training done on PCA data! Time elapsed: {} seconds'.format(time.time()-time_start))

Logistic regression training done on PCA data! Time elapsed: 6.782797336578369 seconds


Question 6: Evaluate the new classifier on the test set: how does it compare to the previous classifier? Discuss the speed / accuracy trade-off and in which case you'd prefer a very slight drop in model performance for a x-time speedup in training.

In [36]:
score = logistic_regression_model.score(X_test_pca, Y_test)
print(score)

0.9232


In [37]:
time_start = time.time()
pca = PCA(n_components=0.99)
X_pca = pca.fit_transform(X)
X_train_pca = X_pca[:60000]
X_test_pca = X_pca[60000:]
logistic_regression_model = LogisticRegression(solver='lbfgs', max_iter=5000, random_state=42, verbose=1)
logistic_regression_model.fit(X_train_pca, Y_train)
print('PCA analysis done! Time elapsed: {} seconds'.format(time.time()-time_start))
score = logistic_regression_model.score(X_test_pca, Y_test)
print(score)

PCA analysis done! Time elapsed: 9.865088701248169 seconds
0.9247


Question 7. Create a new text cell in your Notebook: Complete a 50-100 word summary (or short description of your thinking in applying this week's learning to the solution) of your experience in this assignment. Include: What was your incoming experience with this model, if any? what steps you took, what obstacles you encountered. how you link this exercise to real-world, machine learning problem-solving. (What steps were missing? What else do you need to learn?) This summary allows your instructor to know how you are doing and allot points for your effort in thinking and planning, and making connections to real-world work.

It is clear that pre-processing a large dataset with 284 columns (p=284) and 70K samples (N) can help model performance at very little cost to model accuracy. I also did a PCA with num_components=0.99 and that did not improve accuracy much and added another 50% to the cost to train.

The numbers:

|          | No PCA   | PCA95+train  | PCA99+train  | 
|----------|----------|------------- |------------- |
|   Time   |   32.46  |   6.20       |   9.27       |
|   Score  |  0.9256  |  0.9232      |  0.9247      |


_NOTE_: I also discovered that not scaling the data first had the most detrimental impact on performance.
When the data was not scaled (i.e. I didn't perform: X = (X/255).astype('float32')) which is done on line 11 in the first cell, model performance was much worse. For the unscaled data case here are the numbers:

|          | No PCA   | PCA95+train  | PCA99+train  | 
|----------|----------|------------- |------------- |
|   Time   |  217.86  |   88.75      |   121.39     |
|   Score  |  0.92    |  0.9233      |  0.921*      |

(*) The model failed to converge and accuracy dropped relative to PCA95